### ISSUE TITLE
Crate Addition Request: vmm-reference

### ISSUE THREAD

**Issue by andreeaflorescu** [rust-vmm/community#83]
## Crate Name: vmm-reference

I am not good with names :(
Suggestions welcome.

## Short Description

This repository is the starting point for the rust-vmm VMM Reference Implementation. This crate will create a minimal VMM which can be extended in the future.

## Why is this crate relevant to the rust-vmm project?

This crate serves two purposes:
1. Test the integration of various rust-vmm crates. Right now we don't have any integration tests to account for the interactions between components. The only way the interactions between components are tested right now is by projects using rust-vmm (i.e. Cloud Hypervisor, Firecracker, and the [PoC](https://github.com/petrutlucian94/crosvm) developed by CloudBase)
2. Provide a starting point for people wanting to understand how can rust-vmm be used in practice.

**Comment by michael2012z** [comment]
Hi, everybody,

I think there are more reason to create this new repository:
- Now the construction of vm-virtio and vm-device crates are more prioritized, but this reference implementation repository also need time to grow. We can do the jobs in parallel, don't have to wait until every other crates are ready.
- In kvm-ioctl repository, we begin to see some unit test cases were becoming complex, we'd better have integration tests to build even more complicated scenario.

I have a proposal to follow up @andreeaflorescu 's idea. 
Please take a look.

## Crate Name

How about rust-vmm-glue?

## Short Description
The reference implementation bears 2 major targets (echoing the two purposes mentioned in the beginning of this issue):
* An example to show how to use Rust-VMM.
* Rust-VMM components integration test.

### Design
#### Principles of the design

- Simple
  - The architecture of the reference implementation should be simple. It should be easy for people to understand how rust-vmm components are connected to build a hypervisor. 
  - A good reference is kvmtool (https://git.kernel.org/pub/scm/linux/kernel/git/will/kvmtool.git), it is "a clean, from-scratch, lightweight KVM host tool". This reference shares the same goal. 
  - To keep simple, the reference program is not designed in centerlization mode: no daemon, no API server.

- General purposed
  - The main consuming projects of rust-vmm (Firecracker and Clould-Hypervisor now, CrosVM may in future) each has its own character and focus. But this reference implementation should be general purposed, it should cover variant use cases. For example, it should support devices commonly used as many as possible.
  - Integrate all functional components (crates).

#### Practical Design
- Prototype
  - I am making a prototype of the design at: https://github.com/michael2012z/rust-vmm-glue
- Binary
  - The reference implementation was built into a single binary file. I assume the binary is named as "glue" (means glue of rust-vmm components) in following description.

- Subcommands
  - The reference implementation works in the way of subcommands. The command format looks like:

  - > ``` glue [FLAGS] [subcommand] [OPTIONS] ```

- Process diagram of RUN subcommand
  - ![](https://raw.githubusercontent.com/michael2012z/rust-vmm-glue/master/docs/images/cmd_run.png "RUN subcommand")

- Usage examples

HELP subcommand
```
$ ./target/debug/glue --help
glue 0.1
Rust-VMM based hypervisor.

USAGE:
    glue [FLAGS] [SUBCOMMAND]

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information
    -v, --verbose    Sets the level of verbosity

SUBCOMMANDS:
    help      Prints this message or the help of the given subcommand(s)
    pause     Pause the virtual machine
    resume    Resume the virtual machine
    run       Start the virtual machine
    stop      Stop the virtual machine
    list      List all virtual machines

```

RUN subcommand
```
$ ./target/debug/glue run --help
glue-run 
Start the virtual machine

USAGE:
    glue run [OPTIONS]

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

OPTIONS:
    -c, --cpus <cpus>        Number of CPUs [default: 1]
    -d, --disk <FILE>        Disk image
    -k, --kernel <FILE>      Kernel to boot
    -m, --mem <mem>          Memory size in MB [default: 512]
    -n, --name <name>        A name for the VM
    -p, --params <params>    Kernel command line arguments
```

PAUSE subcommand
```
$ ./target/debug/glue pause --help
glue-pause 
Pause the virtual machine

USAGE:
    glue pause [OPTIONS]

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

OPTIONS:
    -n, --name <name>    Name of the VM
```

### Integration test

#### Cover as many use cases as possible
- Build - Customize hypervisor (glue) by specifying testing manifest file (it's Cargo.toml by default) with different features/components and build. 
  - >  ``` cargo build --manifest-path ./tests/build/example.toml ``` 

- Basic - Basic functions of VM.
  - Lifecycle
  - Kernel load
  - Kernel command line
  - RootFS load
  - vcpus
  - Memory
  - ...
  
- Devices - Device integration tests.
  - PCI
  - GPU
  - Combinations
  - ...
  
- Features - Features of rust-vmm components and their combinations.
  - To be scoped.
  
- Performance - Performance tests, typically:
  - Boot time
  - Process start time
  - Device IO time
  - ...

#### Test framework candidates
This is to be discussed. Should we use Pytest like Firecracker do or use Rust test?
- Pytest (I prefer this)
  - "Fixture" is a powerful feature of Pytest, which makes it easy to prepare a VM instance with proper capability (like netowrk, network, etc) for test case.
  - Firecraker has a good test codebase writen in Pytest for reference.
  - Python is easy to use and to integrate in CI.
- Rust
  - Using Rust native test is an alternative solution. 
  - Align the test code with function code looks good.
  - We need to implement a test framework from scratch. I didn't know much about popular Rust test framework. If there is, things could be different.

**Comment by sameo** [comment]
Maybe a good starting point: https://github.com/CleverCloud/simple-vmm

**Comment by andreeaflorescu** [comment]
@michael2012z thanks for the proposal.

At this point I think people are aligned with having a rust-vmm reference implementation, so we can create the repository. We just need to align on the name.

I very much prefer for it to be called rust-vmm reference implementation and not glue because I am dreaming big:
- this repository should be used as a starting point to create specialized VMMs; this means that people can either fork it (or maybe we have some nice compile time options) to build their own VMM; `glue` is not appropriate for that vision;

I personally dislike the python fixtures; it has complicated the integration test framework to a point where it is very hard to use; it might be that there are better usage patterns out there that make the test code more readable and easier to maintain; if the test infrastructure of the vmm-reference-impl is hard, then nobody will maintain and it will eventually rot; unlike common rust vmm components, people don't have a real buy in to maintain it.

We can discuss further design details in the dedicated repository so that we don't loose them with time. WDYT?

**Comment by michael2012z** [comment]
Hi, @andreeaflorescu 

It's really nice to have more discussion on this topic.

"**be used as a starting point to create specialized VMMs**" is a wonderful idea. The reference implementation will be more useful if people can use it to build VMM. We can design some mechanism to customizing or plugging-in to achieve this. The idea of "glue" is more like a demo to show how to integrate every components together. And I haven't any preference on the name.

Test of the reference implementation is important, it is the integration test for all Rust-VMM components. I agree the test framework need to be easy to use. It's worthy to explore a better framework.

Let's discuss more design in the up-coming repository.

**Comment by aghecenco** [comment]
The [`vmm-reference`](https://github.com/rust-vmm/vmm-reference) crate is up - @michael2012z let's pick up the design discussions here: https://github.com/rust-vmm/vmm-reference/pull/1
